container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java

```diff
@@ -186,70 +189,101 @@ public int groupsWithSufficientCoverage() {
     @Override
     public void failed(Node node) {
         node.setWorking(false);
-
-        // Take ourselves out if we usually dispatch only to our own host
-        if (usesDirectDispatchTo(node))
-            vipStatus.removeFromRotation(clusterId);
+        updateVipStatusOnNodeChange(node, true);
     }
 
     private void updateSufficientCoverage(Group group, boolean sufficientCoverage) {
-        // update VIP status if we direct dispatch to this group and coverage status changed
-        boolean isInRotation = vipStatus.isInRotation();
-        boolean hasChanged = sufficientCoverage != group.hasSufficientCoverage();
-        boolean isDirectDispatchGroupAndChange = usesDirectDispatchTo(group) && hasChanged;
+        if (sufficientCoverage == group.hasSufficientCoverage()) return; // no change
+
         group.setHasSufficientCoverage(sufficientCoverage);
-        if ((!isInRotation || isDirectDispatchGroupAndChange) && sufficientCoverage) {
-            // We will set this cluster in rotation if
-            // - not already in rotation and one group has sufficient coverage.
-            vipStatus.addToRotation(clusterId);
-        } else if (isDirectDispatchGroupAndChange) {
-            // We will take it out of rotation if the group is mandatory (direct dispatch to this group)
-            vipStatus.removeFromRotation(clusterId);
+        updateVipStatusOnCoverageChange(group, sufficientCoverage);
+    }
+
+    private void updateVipStatusOnNodeChange(Node node, boolean working) {
+        if (usesLocalCorpusIn(node)) { // follow the status of the local corpus
+            if (working)
+                vipStatus.addToRotation(clusterId);
+            else
+                vipStatus.removeFromRotation(clusterId);
+        }
+        else {
+            if ( ! hasInformationAboutAllNodes()) return;
+
+            if (hasWorkingNodesWithDocumentsOnline())
+                vipStatus.addToRotation(clusterId);
+            else
+                vipStatus.removeFromRotation(clusterId);
         }
     }
 
-    private boolean usesDirectDispatchTo(Node node) {
-        return directDispatchTarget.isPresent() && directDispatchTarget.get().equals(node);
+    private void updateVipStatusOnCoverageChange(Group group, boolean sufficientCoverage) {
+        boolean isInRotation = vipStatus.isInRotation();
+        if (usesLocalCorpusIn(group)) { // follow the status of the local corpus
+            if (sufficientCoverage)
+                vipStatus.addToRotation(clusterId);
+            else
+                vipStatus.removeFromRotation(clusterId);
+        }
+        else {
+            if ( ! isInRotation && sufficientCoverage)
+                vipStatus.addToRotation(clusterId);
+        }
+    }
+
+    private boolean hasInformationAboutAllNodes() {
+        return nodesByHost.values().stream().allMatch(Node::getStatusIsKnown);
+    }
+
+    private boolean hasWorkingNodesWithDocumentsOnline() {
```

Comment:
I do not understand why we require documents to be present. I think that hasWorkingNodes should do just fine.
There is much more information in 0 documents than cluster is down.

Reply:
It's what we used to do, but yes I can't really think of a good reason either, probably we used to do it because before this code we also used to do it. I'll change it.